pkgs/build-support/build-fhs-userenv/chrootenv.c

```diff
@@ -20,56 +21,112 @@
 #include <sys/stat.h>
 #include <sys/wait.h>
 
-char *env_whitelist[] = {"TERM",
-                         "DISPLAY",
-                         "XAUTHORITY",
-                         "HOME",
-                         "XDG_RUNTIME_DIR",
-                         "LANG",
-                         "SSL_CERT_FILE",
-                         "DBUS_SESSION_BUS_ADDRESS"};
-
-char **env_build(char *names[], size_t len) {
-  char *env, **ret = malloc((len + 1) * sizeof(char *)), **ptr = ret;
-
-  for (size_t i = 0; i < len; i++) {
-    if ((env = getenv(names[i]))) {
-      if (asprintf(ptr++, "%s=%s", names[i], env) < 0)
-        errorf(EX_OSERR, "asprintf");
+#define LEN(x) sizeof(x) / sizeof(*x)
+
+char *env_blacklist[] = {};
+
+char **env_filter(char *envp[]) {
+  char **filtered_envp = malloc(sizeof(*envp));
+  size_t n = 0;
+
+  while (*envp != NULL) {
+    bool blacklisted = false;
+
+    for (size_t i = 0; i < LEN(env_blacklist); i++) {
+      if (!strncmp(*envp, env_blacklist[i], strlen(env_blacklist[i]))) {
+        blacklisted = true;
+        break;
+      }
+    }
+
+    if (!blacklisted) {
+      filtered_envp = realloc(filtered_envp, (n + 2) * sizeof(*envp));
+
+      if (filtered_envp == NULL)
+        errorf(EX_OSERR, "realloc");
+
+      filtered_envp[n++] = *envp;
     }
+
+    envp++;
+  }
+
+  filtered_envp[n] = NULL;
+  return filtered_envp;
+}
+
+void bind(char *from, char *to) {
+  if (mkdir(to, 0755) < 0)
+    errorf(EX_IOERR, "mkdir");
+
+  if (mount(from, to, "bind", MS_BIND | MS_REC, NULL) < 0)
+    errorf(EX_OSERR, "mount");
+}
+
+char *strjoin(char *dir, char *name) {
+  char *path = malloc(strlen(dir) + strlen(name) + 1);
+
+  if (path == NULL)
+    errorf(EX_OSERR, "malloc");
+
+  if (strcpy(path, dir) < 0)
+    errorf(EX_IOERR, "strcpy");
+
+  if (strcat(path, name) < 0)
+    errorf(EX_IOERR, "strcat");
+
+  return path;
+}
+
+char *bind_blacklist[] = {".", "..", "bin", "etc", "host", "usr"};
+
+bool bind_blacklisted(char *name) {
+  for (size_t i = 0; i < LEN(bind_blacklist); i++) {
+    if (!strcmp(bind_blacklist[i], name))
+      return true;
   }
 
-  *ptr = NULL;
-  return ret;
+  return false;
+}
+
+bool isdir(char *path) {
+  struct stat buf;
+  stat(path, &buf);
```

Comment:
Should check return value of `stat` -> buf could even contain the result of previous values stat calls.

Reply:
You are right, I'll reimplement this on top of `opendir`.